File: Sudoku_solver.cpp
class Solution {
private:
    int row[9][256], col[9][256], cube[3][3][256];
public:
    void solveSudoku(vector<vector<char>>& board) {
        memset(row,0,sizeof(row));
        memset(col,0,sizeof(col));
        memset(cube,0,sizeof(col));
        
        // hash the already existing cell values
        for (int r = 0; r < 9; r++) {
            for (int c = 0; c < 9; c++) {
                if (board[r][c] != '.') {
                    int d = board[r][c];
                    row[r][d] = 1; col[c][d] = 1; cube[r/3][c/3][d] = 1;
                }
            }
        }
        
        solveSudoku(board, 0, 0);
    }
private:
    bool check(vector<vector<char>>& board, int r, int c, char val)
    {
        if(row[r][val] == 1) return false;
        if(col[c][val] == 1) return false;
        if(cube[r/3][c/3][val] == 1) return false;
        return true;
    }
    bool solveSudoku(vector<vector<char>> &board, int i, int j)
    {
        if(i==9) return true;
        if(j==9) return solveSudoku(board, i+1, 0);
        if(board[i][j] != '.') return solveSudoku(board, i, j+1);
    
        for(char d='1'; d<='9'; d++)
        {
            if(check(board, i, j, d))
            {
                board[i][j] = d;
                row[i][d] = 1; col[j][d] = 1; cube[i/3][j/3][d] = 1;  // hash the digit 'd'
                if(solveSudoku(board, i, j+1)) return true;
                board[i][j] = '.';
                row[i][d] = 0; col[j][d] = 0; cube[i/3][j/3][d] = 0;  // unhash the digit 'd'
            }
        }
    
        return false;
    }
};

File: Sudoku_Solver.py
# problem : Sudoku Solver
# hacktoberfest 2021

class Solution:
    class _State(object):
        __slot__ = ("board", "row", "col", "usedRow", "usedCol", "usedBlock")
        def __init__(self, board, row, col, usedRow, usedCol, usedBlock):
            self.board = [line*1 for line in board] 
            self.row = row                          
            self.col = col                          
            self.usedRow = usedRow          
            self.usedCol = usedCol          
            self.usedBlock = usedBlock      
    def solveSudoku(self, board):
        # Convert the input (array of strings) to 2D array of integer
        char2int = {"1":1, "2":2, "3":4, "4":8, "5":16, "6":32, "7":64, "8":128, "9":256}
        boardMatrix = [[char2int.get(item, 0) for item in line] for line in board]
        stack = []
        usedRow, usedCol, usedBlock = [0] * 9, [0] * 9, []
        # Get the initially appeared integers in each row and column
        for i in xrange(9):
            usedRow[i] = sum(boardMatrix[i])
            usedCol[i] = sum([line[i] for line in boardMatrix])
        # Get the initially appeared integers in each block
        for iBlock in xrange(3):
            for jBlock in xrange(3):
                usedBlock.append(0)
                for i in xrange(3):
                    for j in xrange(3):
                        usedBlock[-1] = usedBlock[-1] | boardMatrix[iBlock*3+i][jBlock*3+j]
        stack.append(self._State(boardMatrix, 0, 0, usedRow, usedCol, usedBlock))
        current = None
        # Do a DFS
        while len(stack) != 0:
            current = stack.pop()
            while current.row != 9 and current.board[current.row][current.col] != 0:
                current.col += 1
                if current.col == 9:
                    current.row += 1
                    current.col = 0
            if current.row == 9:                
              break
            for toAdd in char2int.values():
                if toAdd & current.usedRow[current.row] != 0:        continue
                if toAdd & current.usedCol[current.col] != 0:        continue
                if toAdd & current.usedBlock[(current.row//3)*3+current.col//3] != 0: continue
                current.board[current.row][current.col] = toAdd
                newUsedRow = current.usedRow * 1
                newUsedRow[current.row] |= toAdd
                newUsedCol = current.usedCol * 1
                newUsedCol[current.col] |= toAdd
                newUsedBlock = current.usedBlock * 1
                newUsedBlock[(current.row//3)*3+current.col//3] |= toAdd
                stack.append(self._State(current.board, current.row, current.col,
                                         newUsedRow, newUsedCol, newUsedBlock))
        del board[:]
        int2char = {0:".", 1:"1", 2:"2", 4:"3", 8:"4", 16:"5", 32:"6", 64:"7", 128:"8", 256:"9"}
        result = ["".join([int2char[i] for i in line]) for line in current.board]
        for line in result:  board.append(line)
        return
